### ISSUE TITLE
When network interface isn't specified, task processes fail to connect to driver.

### ISSUE THREAD

**Issue by aaron276h** [horovod/horovod#1969]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): N/A
2. Framework version: N/A
3. Horovod version: 0.19.2
4. MPI version: N/A (Using Gloo Controller)
5. CUDA version: 10.0
6. NCCL version: 2.6.4
7. Python version: 3.6.10
8. OS and version: Ubuntu 18.04
9. GCC version: 4.8

**Bug report:**

When running multi-machine and not specifying a network interface, task processes fail to connect to driver. Note: using a GLOO controller.

Stack trace:

```
root@ip-172-31-37-52:/# horovodrun -np 2 -H localhost:1,172.31.35.37:1 -p 12345 --verbose ls
Filtering local host names.
Remote host found: 172.31.35.37
Checking ssh on all remote hosts.
SSH was successful into all the remote hosts.
Testing interfaces on all the hosts.
Launched horovod server.
Attempted to launch horovod task servers.
Waiting for the hosts to acknowledge.
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 37436)
Exception happened during processing of request from ('172.31.37.52', 52632)
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 52636)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
Traceback (most recent call last):
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 55358)
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 52642)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
Exception happened during processing of request from ('127.0.0.1', 37442)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
struct.error: unpack requires a buffer of 4 bytes
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 55364)
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 37448)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
----------------------------------------
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.37.52', 55368)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.35.37', 39790)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
----------------------------------------
Exception happened during processing of request from ('172.31.35.37', 39796)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
----------------------------------------
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
Exception happened during processing of request from ('172.31.35.37', 39798)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/socketserver.py", line 654, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/conda/lib/python3.6/socketserver.py", line 364, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/conda/lib/python3.6/socketserver.py", line 724, in __init__
    self.handle()
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 105, in handle
    req = server._wire.read(self.rfile)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 79, in read
    message_len = struct.unpack('i', rfile.read(4))[0]
struct.error: unpack requires a buffer of 4 bytes
----------------------------------------
Launching horovod task function was not successful:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/task_fn.py", line 67, in <module>
    _task_fn(index, driver_addresses, settings)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/task_fn.py", line 27, in _task_fn
    driver_addresses, settings.key, settings.verbose)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/driver/driver_service.py", line 44, in __init__
    match_intf=match_intf)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/service/driver_service.py", line 159, in __init__
    match_intf=match_intf)
  File "/opt/conda/lib/python3.6/site-packages/horovod/run/common/util/network.py", line 172, in __init__
    'Linux.'.format(service_name=service_name, addresses=addresses))
horovod.run.common.util.network.NoValidAddressesFound: Horovod was unable to connect to horovod driver service on any of the following addresses: {'lo': [('127.0.0.1', 4548)], 'ens3': [('172.31.37.52', 4548)], 'docker0': [('172.17.0.1', 4548)]}.

One possible cause of this problem is that horovod currently requires every host to have at least one routable network interface with the same name across all of the hosts. You can run "ifconfig -a" on every host and check for the common routable interface. To fix the problem, you can rename interfaces on Linux.
```

When a network interface is specified, everything works as expected:
```
root@ip-172-31-37-52:/# horovodrun -np 2 -H localhost:1,172.31.35.37:1 -p 12345 --network-interface ens3 ls
Wed May 20 19:01:46 2020[0]<stdout>:bin
Wed May 20 19:01:46 2020[0]<stdout>:boot
.....
```

**Comment by tgaddair** [comment]
Thanks for reporting @aaron276h.

@abditag2 this is the same issue we've been seeing.

**Comment by aaron276h** [comment]
@tgaddair The issue seems related to `settings.key` being omitted from serialization when starting up tasks. Opened a PR with a potential fix.

**Comment by EnricoMi** [comment]
There is a reason why this has been removed from serialisation. Let me check if there is any other way.

**Comment by tgaddair** [comment]
I'm wondering why this error doesn't surface during out integration tests.  I also haven't been able to reproduce it on my local machine (only running under certain conditions in a Docker container).  @aaron276h, does this error always happen in your environment?

**Comment by EnricoMi** [comment]
Gloo does not ssh to run the worker but runs the commadn directly when they all are on the same machine. This might be the reason.

**Comment by aaron276h** [comment]
@tgaddair Yeah this happens every time we run multi-machine without specifying the network interface.

**Comment by EnricoMi** [comment]
would be nice to have an integration test that is truly distributed, i.e. multiple docker instances building a Horovod cluster

**Comment by ShindongLee** [comment]
> Gloo does not ssh to run the worker but runs the command directly when they are on the machine of the driver: https://github.com/horovod/horovod/blob/master/horovod/run/gloo_run.py#L139. This might be the reason.
> 
> I will add a unit test to reproduce and see if I can fix this without adding the key to serialisation. I will look into that tomorrow.

Hi, I experienced the same issue with openmpi so it may not be a gloo's problem.

My environments are as follows:

Framework: Pytorch
Framework version: 1.5.0
Horovod version: 0.19.2
MPI version: 4.0.0
CUDA version: 10.2
NCCL version: 2.6.4
Python version: 3.6.3
OS and version: Ubuntu 18.04

When I specify network interface with --network-interface option, the issue's gone.
Hope my info helps you.

**Comment by tgaddair** [comment]
Hey @ShindongLee @aaron276h, we've released [v0.19.3](https://github.com/horovod/horovod/releases/tag/v0.19.3) with a fix for this issue.  Let us know if any additional issues come up.

Thanks @aaron276h and @EnricoMi for diagnosing and fixing the issue.
